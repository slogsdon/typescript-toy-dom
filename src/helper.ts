/**
 * The JSX pragma helper function
 *
 * @module typescript-toy-dom/helper
 */

/**
 * Define module imports/exports
 */
import { INode, IProps, NodeLike, NodeType } from "./types";

/**
 * JSX pragma helper
 *
 * Used when compiling JSX to JS function calls.
 *
 * @param type The desired type of node. See `NodeType`.
 * @param props If present, the defined properties for the node.
 * @param children Any child nodes.
 */
export function h(
  type: NodeType,
  props: IProps | undefined,
  ...children: NodeLike[]
): INode {
  return { children, props: props || {}, type };
}

/**
 * Alias of h
 *
 * @see h
 */
export const createElement = h;
