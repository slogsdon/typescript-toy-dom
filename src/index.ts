/**
 * The main entrypoint for the project.
 *
 * @module typescript-toy-dom
 */

/**
 * Define module imports/exports
 */
export * from "./helper";
export * from "./render";
export * from "./types";
