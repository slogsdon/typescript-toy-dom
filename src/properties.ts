/**
 * Handle component properties
 *
 * @module typescript-toy-dom/properties
 */

/**
 * Define module imports/exports
 */
import { IProps, PropValue } from "./types";

/**
 * Tests if a property name matches to that of a custom property
 *
 * @param name The property name to test.
 */
export function isCustomProp(name: string) {
  return false;
}

/**
 * Removes a boolean property
 *
 * @param el The current DOM node.
 * @param name The boolean property to remove.
 */
export function removeBooleanProp(el: Element, name: string) {
  el.removeAttribute(name);
  (el as any)[name] = false;
}

/**
 * Removes a single property
 *
 * @param el The target DOM node.
 * @param name The property name.
 * @param value The property value.
 */
export function removeProp(el: Element, name: string, value?: PropValue) {
  if (isCustomProp(name)) {
    return;
  }

  if (typeof value === "boolean") {
    return removeBooleanProp(el, name);
  }

  return el.removeAttribute(name === "className" ? "class" : name);
}

/**
 * Sets a boolean property
 *
 * Boolean properties are special. Their attribute is set with their
 * name, while their property value is a true boolean.
 *
 * @param el The current DOM node set.
 * @param name The boolean property.
 * @param value The property's value.
 */
export function setBooleanProp(el: Element, name: string, value: boolean) {
  if (value) {
    el.setAttribute(name, name);
    (el as any)[name] = true;
  } else {
    (el as any)[name] = false;
  }
}

/**
 * Sets a single property
 *
 * @param el The target DOM node.
 * @param name The property name.
 * @param value The property value.
 */
export function setProp(el: Element, name: string, value: PropValue) {
  if (isCustomProp(name)) {
    return;
  }

  if (typeof value === "boolean") {
    return setBooleanProp(el, name, value);
  }

  if (typeof value === "string") {
    return el.setAttribute(name === "className" ? "class" : name, value);
  }
}

/**
 * Sets all property values
 *
 * @param el The target DOM node.
 * @param props The property set.
 */
export function setProps(el: Element, props: IProps) {
  if (!props) {
    return [];
  }

  return Object.keys(props).map((name) => setProp(el, name, props[name]));
}

/**
 * Updates a single property
 *
 * @param el The target DOM node.
 * @param name The property name.
 * @param newValue The new property value.
 * @param oldValue The old property value.
 */
export function updateProp(
  el: Element,
  name: string,
  newValue: PropValue,
  oldValue: PropValue,
) {
  if (!newValue) {
    return removeProp(el, name, oldValue);
  }

  if (!oldValue || newValue !== oldValue) {
    return setProp(el, name, newValue);
  }
}

/**
 * Updates all property values
 *
 * @param el The target DOM node.
 * @param newProps The new property set.
 * @param oldProps The old property set.
 */
export function updateProps(
  el: Element,
  newProps: IProps,
  oldProps: IProps = {},
) {
  const props = (Object as any).assign({}, newProps, oldProps);
  return Object.keys(props).map((name) => {
    return updateProp(el, name, newProps[name], oldProps[name]);
  });
}
