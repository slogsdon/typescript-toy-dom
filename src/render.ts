/**
 * Defines various strategies for rendering nodes
 */

/**
 * Define module imports/exports
 */
import { createElement, NodeLike, updateElement } from "./index";

/**
 * The previously rendered node. Used for diffing.
 */
let previousNode: NodeLike;

/**
 * The previously render target. Used for retargetting on
 * follow up calls to `render`.
 */
let previousTarget: Element;

/**
 * Renders a node to it's native DOM form and appends it to the
 * specified `target`
 *
 * If no `target` is provided, `previousTarget` is used if a value
 * has bee previously provided.
 *
 * NOTE: This will remove all children of `target` prior to
 * appending any created DOM nodes.
 *
 * @param node The current node.
 * @param target The DOM node which should act as the node's
 *               parent. If a `string` is used, the value should
 *               be an appropriate CSS selector compatible with
 *               `document.querySelector`.
 */
export function render(node: NodeLike, target?: string | Element) {
  if (typeof target === "string") {
    target = document.querySelector(target);
  }

  // a missing/null target would throw an error anyways,
  // but we give an explicit reason
  if (!target && !previousTarget) {
    throw new Error("Missing/null render target");
  }

  // react style element insertion is desired, so remove
  // existing children in the target
  while (!previousNode && target.firstChild) {
    target.removeChild(target.firstChild);
  }

  updateElement(target || previousTarget, node, previousNode);

  // update globals to track previous arguments
  previousNode = node;
  previousTarget = target;
}

/**
 * Renders a node to it's static string form
 *
 * @param node The current node.
 */
export function renderToStaticMarkup(node: NodeLike) {
  const el = createElement(node);

  if (el instanceof Text) {
    return el.wholeText;
  }

  return el.outerHTML;
}
