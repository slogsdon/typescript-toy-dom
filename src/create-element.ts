/**
 * DOM manipulation functions
 *
 * @module typescript-toy-dom/create-element
 */

/**
 * Define module imports/exports
 */
import { setProps } from "./properties";
import { NodeLike } from "./types";

/**
 * Creates a raw DOM node from a `node`.
 *
 * @param node The current `NodeLike` from which to create the raw DOM node.
 */
export function createElement(node?: NodeLike): Text | Element {
  if (typeof node === "undefined") {
    return;
  }

  if (typeof node === "string") {
    return document.createTextNode(node);
  }

  if (isArray(node)) {
    return (node as any).map(createElement);
  }

  // allow functions as elements
  if (typeof node.type === "function") {
    node.props.children = node.children;
    const ret = node.type(node.props);
    return createElement(ret);
  }

  // create DOM element
  const el = document.createElement(node.type);

  // set props as DOM element attributes
  setProps(el, node.props);

  // create children
  if (node.children) {
    node.children.map(createElement).forEach(appendTo(el));
  }

  return el;
}

/**
 * Test if a value is an array
 *
 * @param obj Value to test
 */
function isArray(obj: any) {
  return Object.prototype.toString.call(obj) === "[object Array]";
}

/**
 * Wraps `Element.appendChild` to handle array-like children.
 *
 * @param el Parent element
 */
function appendTo(el: Element) {
  return (child: Text | Element) => {
    if (isArray(child)) {
      (child as any).map(appendTo(el));
      return;
    }

    el.appendChild(child);
  };
}
