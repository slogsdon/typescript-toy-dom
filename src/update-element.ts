/**
 * DOM manipulation functions
 *
 * @module typescript-toy-dom/update-element
 */

/**
 * Define module imports/exports
 */
import { createElement } from "./create-element";
import { updateProps } from "./properties";
import { NodeLike } from "./types";

/**
 * Updates a raw DOM node from an `oldNode` to a `newNode`.
 *
 * @param parent The raw DOM node which contains the current node.
 * @param newNode The new `NodeLike` from which to create the raw DOM node.
 * @param oldNode The old `NodeLike` with which to compare the `newNode` value.
 * @param index The node's position in the `parent`'s list of children.
 */
export function updateElement(
  parent: Node,
  newNode?: NodeLike,
  oldNode?: NodeLike,
  index = 0,
) {
  // the raw DOM node represented by `oldNode`
  const currentNode = parent.childNodes[index];

  if (!oldNode || !currentNode) {
    const createdElement = createElement(newNode);

    if (!createdElement) {
      return;
    }

    return parent.appendChild(createdElement);
  }

  if (currentNode && !newNode) {
    return parent.removeChild(currentNode);
  }

  if (currentNode && changed(oldNode, newNode)) {
    return parent.replaceChild(createElement(newNode), currentNode);
  }

  // this state should be caught by one of the above if expressions,
  // but in case it's not caught, return early because strings don't
  // have children.
  if (typeof oldNode === "string" || typeof newNode === "string") {
    return;
  }

  // we only know how to deal with `Text` nodes (which cannot possess
  // properties) and `Element` nodes (which can possess properties),
  // so only attempt to set properties for `Element` nodes.
  if (currentNode && currentNode.nodeType === Node.ELEMENT_NODE) {
    updateProps(
      // make an explicit cast even though the current node has already
      // been checked against the `Element` node type.
      currentNode as Element,
      newNode.props,
      oldNode.props,
    );
  }

  const newLength = newNode.children.length;
  const oldLength = oldNode.children.length;

  // `i` should increment to `max(newLength, oldLength)`
  for (let i = 0; i < newLength || i < oldLength; i++) {
    updateElement(currentNode, newNode.children[i], oldNode.children[i], i);
  }
  return;
}

/**
 * Tests if `node1` and `node2` differ.
 *
 * @param node1 Often the old node representation.
 * @param node2 Often the new node representation.
 */
function changed(node1: NodeLike, node2: NodeLike): boolean {
  return (
    typeof node1 !== typeof node2 ||
    (typeof node1 === "string" && node1 !== node2) ||
    (typeof node1 !== "string" &&
      typeof node2 !== "string" &&
      node1.type !== node2.type)
  );
}
