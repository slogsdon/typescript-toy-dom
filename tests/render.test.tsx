import { h } from "../src/helper";
import { render, renderToStaticMarkup } from "../src/render";

describe("render", () => {
  it("throws an error with missing/null target", () => {
    const expectedMessage = "Missing/null render target";

    expect(() => render("hello", "#missing-element")).toThrowError(
      expectedMessage,
    );
    expect(() => render("hello", null)).toThrowError(expectedMessage);
  });

  it("renders to a target element by reference", () => {
    const target = document.createElement("div");

    render("hello", target);

    expect(target.childNodes).toHaveLength(1);
    expect(target.firstChild).toBeInstanceOf(Text);
    expect(target).toMatchSnapshot();
  });

  it("renders to a target element by selector", () => {
    render("hello", "body");

    expect(document.body.childNodes).toHaveLength(1);
    expect(document.body.firstChild).toBeInstanceOf(Text);
    expect(document.body).toMatchSnapshot();
  });

  it.skip("keeps previous children on render with flag", () => {
    const target = document.createElement("div");
    target.appendChild(document.createTextNode("hello"));

    expect(target.childNodes).toHaveLength(1);
    expect(target.firstChild).toBeInstanceOf(Text);

    render("hello", target, false);

    expect(target.childNodes).toHaveLength(2);
    expect(target.firstChild).toBeInstanceOf(Text);
  });

  describe("strings", () => {
    let parent: Element;

    beforeEach(() => {
      parent = document.createElement("div");
    });

    afterEach(() => {
      parent.innerHTML = "";
      parent = null;
    });

    it("should render Text", () => {
      render("hello", parent);

      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Text);
      expect(parent).toMatchSnapshot();

      render("goodbye", parent);

      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Text);
      expect(parent).toMatchSnapshot();
    });
  });

  describe("functions", () => {
    let parent: Element;

    beforeEach(() => {
      parent = document.createElement("div");
    });

    afterEach(() => {
      parent.innerHTML = "";
      parent = null;
    });

    it("should render nothing when empty", () => {
      const Fn: any = jest.fn(() => undefined);

      render(<Fn />, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(0);
      expect(parent).toMatchSnapshot();

      render(<Fn />, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(0);
      expect(parent).toMatchSnapshot();
    });

    it("should render Element", () => {
      const Fn: any = jest.fn(({ children }) => <div>{children}</div>);

      render(<Fn>hello</Fn>, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Element);
      expect(parent.firstChild.childNodes).toHaveLength(1);
      expect(parent.firstChild.firstChild).toBeInstanceOf(Text);
      expect(parent).toMatchSnapshot();

      render(<Fn>goodbye</Fn>, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Element);
      expect(parent.firstChild.childNodes).toHaveLength(1);
      expect(parent.firstChild.firstChild).toBeInstanceOf(Text);
      expect(parent).toMatchSnapshot();
    });

    it("should render Element with removed children during update", () => {
      const Fn: any = jest.fn(({ children }) => <div>{children}</div>);

      render(<Fn>hello</Fn>, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Element);
      expect(parent.firstChild.childNodes).toHaveLength(1);
      expect(parent.firstChild.firstChild).toBeInstanceOf(Text);
      expect(parent).toMatchSnapshot();

      render(<Fn />, parent);

      expect(Fn).toHaveBeenCalled();
      expect(parent.childNodes).toHaveLength(1);
      expect(parent.firstChild).toBeInstanceOf(Element);
      expect(parent.firstChild.childNodes).toHaveLength(0);
      expect(parent).toMatchSnapshot();
    });
  });
});

describe("renderToStaticMarkup", () => {
  describe("strings", () => {
    it("should render Text", () => {
      const result = renderToStaticMarkup("hello");

      expect(result).toMatchSnapshot();
    });
  });

  describe("functions", () => {
    it("should render nothing when empty", () => {
      const Fn: any = jest.fn(() => undefined);

      const result = renderToStaticMarkup(<Fn />);

      expect(result).toMatchSnapshot();
    });

    it("should render Element", () => {
      const Fn: any = jest.fn(({ children }) => <div>{children}</div>);

      const result = renderToStaticMarkup(<Fn>hello</Fn>);

      expect(Fn).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });
  });
});
