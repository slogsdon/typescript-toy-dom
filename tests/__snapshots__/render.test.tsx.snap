// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render functions should render Element 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`render functions should render Element 2`] = `
<div>
  <div>
    goodbye
  </div>
</div>
`;

exports[
  `render functions should render Element with removed children during update 1`
] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[
  `render functions should render Element with removed children during update 2`
] = `
<div>
  <div />
</div>
`;

exports[`render functions should render nothing when empty 1`] = `<div />`;

exports[`render functions should render nothing when empty 2`] = `<div />`;

exports[`render renders to a target element by reference 1`] = `
<div>
  hello
</div>
`;

exports[`render renders to a target element by selector 1`] = `
<body>
  hello
</body>
`;

exports[`render strings should render Text 1`] = `
<div>
  hello
</div>
`;

exports[`render strings should render Text 2`] = `
<div>
  goodbye
</div>
`;

exports[
  `renderToStaticMarkup functions should render Element 1`
] = `"<div>hello</div>"`;

exports[
  `renderToStaticMarkup functions should render nothing when empty 1`
] = `""`;

exports[`renderToStaticMarkup strings should render Text 1`] = `"hello"`;
